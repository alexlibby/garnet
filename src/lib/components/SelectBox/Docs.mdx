import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as SelectBoxStories from "./SelectBox.stories";

<Meta of={SelectBoxStories} />

# SelectBox

SelectBox is the primary component. It has two possible states.

- [Default](#default)
- [With Label](#withlabel)
- [Disabled](#disabled)
- [Disabled2](#disabled2)

## Default

This is the default state for our SelectBox component:

<Canvas of={SelectBoxStories.Default} />

## With Label

This is the default state for our SelectBox component:

<Canvas of={SelectBoxStories.WithLabel} />

## Disabled

This is the SelectBox component in its disabled state:

<Canvas of={SelectBoxStories.Disabled} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={SelectBoxStories} />
