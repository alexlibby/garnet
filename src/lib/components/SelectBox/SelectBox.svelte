<svelte:options customElement="garnet-selectbox" />

<script>
  import { createEventDispatcher } from 'svelte';

  export let selectOptions = [];
  export let displayText = (a) => a.value;
  export let index = 1;
  export let disabled = false;
  export let selected = {};
  export let label = "Test dropdown:"

  const dispatch = createEventDispatcher();

  $: selected = selectOptions.find((o) => o.id === index);
  
  function handleChange() {
    dispatch('change', { text: selected });
  }
</script>

<div class="garnet-selectbox">
  <label for="garnet-selectbox">
    {label}
    <select bind:value={index} on:change={handleChange} data-testid="selectdropdown" title="garnet-selectbox" {disabled}>
      {#each selectOptions as option, i}
        <option value={i + 1}>
          {displayText(option)}
        </option>
      {/each}
    </select>
  </label>
</div>

<style>
  .garnet-selectbox {
    display: flex;
    align-items: center;
    font-family: Arial, Helvetica, sans-serif;
  }

  label {
    padding-right: 10px;
  }

  select {
    padding: 5px 100px 5px 5px; 
    font-size: 16px;
    border: 1px solid #733635;
    height: 34px;
    border-radius: 10px;
  }
  
  option:disabled {
    cursor: not-allowed;
  }
</style>
