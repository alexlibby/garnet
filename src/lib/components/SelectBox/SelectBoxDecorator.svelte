<div>
  <slot />
</div>
<div id="message4">SelectBox value is (not selected)</div>

<style>
  div {
    margin-top: 10px;
  }

  #message4 {
    font-family: Arial, Helvetica, sans-serif;
  }
</style>