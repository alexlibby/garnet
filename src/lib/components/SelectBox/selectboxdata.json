[
  { "id": 1, "value": "aaa" },
  { "id": 2, "value": "bbb" },
  { "id": 3, "value": "ccc" },
  { "id": 4, "value": "ddd" }
]
