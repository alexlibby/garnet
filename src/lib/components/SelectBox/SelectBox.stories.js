import SelectBox from "./SelectBox.svelte";
import SelectBoxDecorator from "./SelectBoxDecorator.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

let optionValues = [
  { id: 1, value: "aaa" },
  { id: 2, value: "bbb" },
  { id: 3, value: "ccc" },
  { id: 4, value: "ddd" },
];

export default {
  title: "Garnet UI Library/Action Components/SelectBox",
  component: SelectBox,
  decorators: [() => SelectBoxDecorator],
  argTypes: {
    checked: { control: "boolean" },
    label: { control: "value" },
    selectOptions: optionValues,
  },
};

export const Default = () => ({
  Component: SelectBox,
  props: {
    selectOptions: optionValues,
    label: "",
  },
  on: {
    change: (event) => {
      document.getElementById("message4").innerHTML =
        `SelectBox value is ${JSON.stringify(event.detail.text.value)}`;
    },
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const WithLabel = () => ({
  Component: SelectBox,
  props: {
    selectOptions: optionValues,
    label: "This is a test:",
  },
  on: {
    change: (event) => {
      document.getElementById("message4").innerHTML =
        `SelectBox value is ${JSON.stringify(event.detail.text.value)}`;
    },
  },
});

WithLabel.parameters = {
  badges: [BADGES.ALPHA],
};

export const Disabled = () => ({
  Component: SelectBox,
  props: {
    selectOptions: optionValues,
    disabled: true,
  },
});

Disabled.parameters = {
  badges: [BADGES.ALPHA],
};
