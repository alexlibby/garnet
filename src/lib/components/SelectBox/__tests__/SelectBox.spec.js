import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import SelectBox from "../../SelectBox/SelectBox.svelte";

const SelectBoxHTML = `<div class="garnet-selectbox s-YBKB7XKMQi9t"><label for="garnet-selectbox" class="s-YBKB7XKMQi9t"><select data-testid="selectdropdown" title="garnet-selectbox" class="s-YBKB7XKMQi9t"><option value="1" class="s-YBKB7XKMQi9t">aaa </option><option value="2" class="s-YBKB7XKMQi9t">bbb </option><option value="3" class="s-YBKB7XKMQi9t">ccc </option><option value="4" class="s-YBKB7XKMQi9t">ddd </option></select> </label></div>`;
expect.extend(AxeMatchers);

describe("Tests for SelectBox", () => {
  const props = {
    selectOptions: [{ text: "aaa" }, { text: "bbb" }, { text: "ccc" }],
    index: 0,
    disabled: false,
  };

  test("should render properly", () => {
    const result = render(SelectBox, props);
    expect(() => result).not.toThrow();
  });

  test("should select the second option if clicked", async () => {
    const result = render(SelectBox, props);

    await fireEvent.click(screen.getByTestId("selectdropdown"));
    screen.findByText(props.selectOptions[0].text).then(function (result) {
      expect(result.text).toEqual("aaa");
    });
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new SelectBox({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => SelectBoxHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
