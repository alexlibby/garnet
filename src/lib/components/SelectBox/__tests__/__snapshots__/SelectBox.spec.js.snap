// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for SelectBox > get a snapshot of component 1`] = `"<div class="garnet-selectbox svelte-wpck7s"><label for="garnet-selectbox" class="svelte-wpck7s">Test dropdown: <select data-testid="selectdropdown" title="garnet-selectbox" class="svelte-wpck7s"><option value="1" class="svelte-wpck7s">undefined </option><option value="2" class="svelte-wpck7s">undefined </option><option value="3" class="svelte-wpck7s">undefined </option></select></label></div>"`;
